alt_e2eshark/run.py

```diff
@@ -0,0 +1,415 @@
+            if curr_stage in stages:
+                golden_outputs = inst.apply_postprocessing(golden_outputs_raw)
+                outputs = inst.apply_postprocessing(outputs_raw)
+        except Exception as e:
+            log_exception(e, log_dir, curr_stage, t.unique_name, verbose)
+            continue
+
+        # store the results
+        if "setup" and "native_inference" and "compiled_inference" in stages:
+            try:
+                result = TestResult(
+                    name=t.unique_name,
+                    input=inputs,
+                    gold_output=golden_outputs,
+                    output=outputs,
+                )
+                # log the results
+                test_passed = log_result(result, log_dir, [1e-3, 1e-3])
+                num_passes += int(test_passed)
+                if verbose:
+                    to_print = "\tPASS" if test_passed else "\tFAILED (Numerics)"
+                    print(to_print)
+                elif not test_passed:
+                    print(f"FAILED: {t.unique_name}")
+            except Exception as e:
+                log_exception(e, log_dir, "results-summary", t.unique_name, verbose)
+
+    print("\nTest Summary:")
+    print(f"\tPASSES: {num_passes}\n\tTOTAL: {len(test_list)}")
+    print(f"results stored in {parent_log_dir}")
+
+
+def log_result(result, log_dir, tol):
+    # TODO: add more information for the result comparison (e.g., on verbose, add information on where the error is occuring, etc)
+    summary = result_comparison(result, tol)
+    num_match = 0
+    num_total = 0
+    for s in summary:
+        num_match += s.sum().item()
+        num_total += s.nelement()
+    percent_correct = num_match / num_total
+    with open(log_dir + "inference_comparison.log", "w+") as f:
+        f.write(
+            f"matching values with (rtol,atol) = {tol}: {num_match} of {num_total} = {percent_correct*100}%\n"
+        )
+        f.write(f"Test Result:\n{result}")
+    return num_match == num_total
+
+
+def log_exception(e: Exception, path: str, stage: str, name: str, verbose: bool):
+    '''generates a log for an exception generated during a testing stage'''
+    log_filename = path + stage + ".log"
+    base_str = f"Failed test at stage {stage} with exception:\n{e}\n"
+    with open(log_filename, "w") as f:
+        f.write(base_str)
+        if verbose:
+            print(f"\tFAILED ({stage})")
+            import traceback
+
+            traceback.print_exception(e, file=f)
+        else:
+            print(f"FAILED: {name}")
+
+
+def _get_argparse():
+    msg = "The run.py script to run e2e shark tests"
+    parser = argparse.ArgumentParser(prog="run.py", description=msg, epilog="")
+
+    # test config related arguments:
+    parser.add_argument(
+        "-b",
+        "--backend",
+        choices=["llvm-cpu", "amd-aie", "rocm", "vulkan"],
+        default="llvm-cpu",
+        help="Target backend i.e. hardware to run on",
+    )
```

Comment:
This should either include both
1. a compiler target backend 
2. a runtime HAL device

_or_ it should collect both of those into a single configuration

For examples of configurations, see
* https://github.com/iree-org/iree/tree/main/build_tools/pkgci/external_test_suite
* https://github.com/iree-org/iree/blob/main/build_tools/python/e2e_test_framework/definitions/iree_definitions.py
* https://github.com/iree-org/iree/blob/0934526e288315552650554da12d49b94845c7bf/integrations/tensorflow/python_projects/iree_tf/iree/tf/support/module_utils.py#L961-L990